### ISSUE TITLE
Make parsers more easily pluggable

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#4640]
I've been looking at some of the hoops we jump through (and `babel-eslint` jumps through) to get parsers to work in ESLint. It seems like we could dramatically improve things with just a few simple steps. Here's what I'm proposing:

1. Parsers should export a `Syntax` object that contains the names of all AST nodes they support. Esprima and Espree already do this.
1. Parsers should also export a `VisitorKeys` object containing the traversal information for the nodes they support. estraverse exports this currently.

When a custom parser is specified, ESLint will augment the `Syntax` and `VisitorKeys` estraverse objects with those that are exported from the parser itself, ensuring ESLint can work properly.

The benefits, as far as I can see them, are:

1. Custom parsers won't have to monkey patch estraverse to get traversal to work correctly (as `babel-eslint` does today).
1. The syntax information now lives in one place (the parser) rather than being split between the parser and ESLint. That means we can add more to Espree and ESLint will Just Work (tm).

This would also benefit `typescript-eslint-parser`, which is sure to have some nonstandard nodes to deal with.

Thoughts?

@hzoo

**Comment by hzoo** [comment]
That would certainly help with patching up the visitorKeys - so instead of monkeypatching it in babel-eslint, it would export the visitorKeys and eslint would patch it instead? This would help to remove https://github.com/babel/babel-eslint/blob/master/index.js#L44-L62

**Comment by nzakas** [comment]
Yup! Exactly.

**Comment by hzoo** [comment]
Don't know if there is an easy way (or if you've thought of) doing the same for escope? There is a lot of patching done for flow, experimental features regarding `no-unused-vars` for that.

**Comment by nzakas** [comment]
I haven't done much thinking around that, as it seemed like traversal was the lowest hanging fruit. Do you have a suggestion?

**Comment by hzoo** [comment]
It's a similar thing but instead we have to patch the escope visitor methods somehow to also visit the nodes it doesn't know about (example with decorators)

```js
function visitDecorators(node) {
    if (!node.decorators) {
      return;
    }
    for (var i = 0; i < node.decorators.length; i++) {
      if (node.decorators[i].expression) {
        this.visit(node.decorators[i]);
      }
    }
  }
```

**Comment by nzakas** [comment]
It looks like you're doing more than just adding methods to the referencer, though. I'm not sure there's an easy way to do this generically.

**Comment by hzoo** [comment]
Most of it is either visiting certain nodes by rewriting the function to do so and then calling the original referencer methods or creating a scope variable if needed. Maybe this is more for escope

**Comment by nzakas** [comment]
Yeah, it seems like more of an escope thing. Though, I wonder if we could go through and use `VisitorKeys` to create default visitors for things to at least get part of the way there. We can keep thinking about it going forward, though I would like to see escope become more extensible so we're not reaching into files and hacking things in.

**Comment by nzakas** [comment]
Closing in favor of https://github.com/eslint/eslint/issues/5476
